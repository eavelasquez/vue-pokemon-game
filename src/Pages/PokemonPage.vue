<template>
  <h1 v-if="!pokemon">Wait a moment please...</h1>
  <div v-else>
    <h1>Who is this Pokemon?</h1>
    <PokemonPicture :pokemon-id="pokemon.id" :show-pokemon="showPokemon" />
    <PokemonOptions :pokemons="pokemons" @selection-pokemon="checkAnswer" />

    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">New Game</button>
    </template>
  </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'

import getPokemonOptions from '@/helpers/getPokemonOptions.js'

export default {
  name: 'PokemonPage',
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      message: '',
      pokemon: null,
      pokemons: [],
      showAnswer: false,
      showPokemon: false,
    }
  },
  methods: {
    async mixPokemonOptions() {
      this.pokemons = await getPokemonOptions()

      const randomInt = Math.floor(Math.random() * 4)
      this.pokemon = this.pokemons[randomInt]
    },
    checkAnswer(selectedId) {
      this.showAnswer = true
      this.showPokemon = true

      this.message =
        selectedId === this.pokemon.id
          ? `You're right, this pokemon is ${this.pokemon.name}!`
          : `Ops! You're wrong, this pokemon is ${this.pokemon.name}`
    },
    newGame() {
      (this.message = ''),
        (this.pokemon = null),
        (this.showAnswer = false),
        (this.showPokemon = false),
        (this.pokemons = []),
        this.mixPokemonOptions()
    },
  },
  mounted() {
    this.mixPokemonOptions()
  },
}
</script>

<style scoped>
button {
  background-color: #64bb87;
  border-radius: 5px;
  border: 1px solid white;
  color: white;
  cursor: pointer;
  margin: 5px 5px;
  padding: 5px 15px;
  transition: 0.3s ease-in-out;
}

button:hover {
  background-color: #5aa67b;
  transition: 0.3s ease-in-out;
}
</style>