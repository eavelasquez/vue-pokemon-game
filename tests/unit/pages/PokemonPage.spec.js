import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage.vue'

import { pokemons } from '../mocks/pokemons.mock'


describe('PokemonPage.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage)
  })

  test('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should match snapshot when pokemons load', () => {
    const wrapper = mount(PokemonPage, {
      data() {
        return {
          message: '',
          pokemon: pokemons[0],
          pokemons: pokemons,
          showAnswer: false,
          showPokemon: false,
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should call mixPokemonOptions when mounting component', () => {
    const mixPokemonOptionsSpy = jest.spyOn(PokemonPage.methods, 'mixPokemonOptions')
    shallowMount(PokemonPage)

    expect(mixPokemonOptionsSpy).toHaveBeenCalled()
  })

  test('should show the PokemonPicture and PokemonOptions components', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          message: '',
          pokemon: pokemons[0],
          pokemons: pokemons,
          showAnswer: false,
          showPokemon: false,
        }
      }
    })

    const picture = wrapper.findComponent({ name: 'pokemon-picture' })
    const options = wrapper.findComponent({ name: 'pokemon-options' })

    expect(picture.exists()).toBeTruthy()
    expect(picture.attributes('pokemonid')).toBe('1')
    expect(options.exists()).toBeTruthy()
    expect(options.attributes('pokemons')).toBeTruthy()
  })

  test('checkAnswer should work', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          message: '',
          pokemon: pokemons[1],
          pokemons: pokemons,
          showAnswer: false,
          showPokemon: false,
        }
      }
    })

    await wrapper.vm.checkAnswer(pokemons[1].id)

    expect(wrapper.find('h2').exists()).toBeTruthy()
    expect(wrapper.vm.showPokemon).toBeTruthy()
    expect(wrapper.vm.showAnswer).toBeTruthy()
    expect(wrapper.vm.message).toBe(`You\'re right, this pokemon is ${pokemons[1].name}!`)

    await wrapper.vm.checkAnswer(pokemons[2].id)
    expect(wrapper.vm.message).toBe(`Ops! You're wrong, this pokemon is ${pokemons[1].name}`)
  })
  
})