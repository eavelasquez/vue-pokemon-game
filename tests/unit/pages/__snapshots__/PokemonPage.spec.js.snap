// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage.vue should match snapshot 1`] = `<h1>Wait a moment please...</h1>`;

exports[`PokemonPage.vue should match snapshot when pokemons load 1`] = `
<div>
  <h1>Who is this Pokemon?</h1>
  <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg" class="hidden-pokemon" alt="pokemon"></div>
  <div class="options-container">
    <ul>
      <li>bulbasaur</li>
      <li>ivysaur</li>
      <li>venusaur</li>
      <li>charmander</li>
    </ul>
  </div>
  <!--v-if-->
</div>
`;
